rcl/include/rcl/time.h

```diff
@@ -54,12 +54,67 @@ typedef enum rcl_clock_type_t
   RCL_STEADY_TIME
 } rcl_clock_type_t;
 
+/// A duration of time, measured in nanoseconds and its source.
+typedef struct rcl_duration_t
+{
+  rcl_duration_value_t nanoseconds;
+} rcl_duration_t;
+
+/// Enumeration to describe the type of time jump.
+typedef enum rcl_clock_change_t
+{
+  /// The source before and after the jump is ROS_TIME.
+  RCL_ROS_TIME_NO_CHANGE = 1,
+  /// The source switched to ROS_TIME from SYSTEM_TIME.
+  RCL_ROS_TIME_ACTIVATED = 2,
+  /// The source switched to SYSTEM_TIME from ROS_TIME.
+  RCL_ROS_TIME_DEACTIVATED = 3,
+  /// The source before and after the jump is SYSTEM_TIME.
+  RCL_SYSTEM_TIME_NO_CHANGE = 4
+} rcl_clock_change_t;
+
+/// Struct to describe a jump in time.
+typedef struct rcl_time_jump_t
+{
+  /// Indicate whether or not the source of time changed.
+  rcl_clock_change_t clock_change;
+  /// The new time minus the last time before the jump.
+  rcl_duration_t delta;
+} rcl_time_jump_t;
+
+/// Signature of a time jump callback.
+typedef void (* rcl_jump_callback_t)(
+  const struct rcl_time_jump_t * time_jump,
+  bool before_jump,
+  void * user_data);
+
+/// Describe the prerequisites for calling a time jump callback.
+typedef struct rcl_jump_threshold_t
+{
+  /// True to call callback when the clock type changes.
+  bool on_clock_change;
+  /// Minimum jump forwards to be considered exceeded, or zero to disable.
+  rcl_duration_t min_forward;
+  /// Minimum jump backwards to be considered exceeded, or zero to disable.
```

Comment:
962c0edee9c039b6b051c0a8c2a6e1b142b6395d

Reply:
Sounds good to me. Should the same semantic be applied to the forward jump duration?

Reply:
>Should the same semantic be applied to the forward jump duration?

Referring to bbc55a5? I had done that, but reverted before committing because of the opposite case: someone might think to put a negative duration to disable the callback and it would be silently accepted.

Reply:
I just don't like the asymmetry of it. On the one side (backward) we want users to specify positive instead of negative durations which are still valid  but on the other (forward) we want to interpret the opposite sign as disabled...

Reply:
> I just don't like the asymmetry of it

How about 2c9bbc8e4307553d3676dae3c69215e6dd6ee48f? I'm out of ideas for how else to get the best of both worlds. For symmetry it requires both `min_backward` and `min_forward` to be specified as positive magnitudes. Both are disallowed from being negative rather than disabled to avoid silently missing jumps if a user mistakenly sets a negative backwards jump.

Reply:
Well, I just find it weird that `min_forward` is picky about the sign of the duration while `min_backward` is not (always using abs and interpreting it as a backward duration).

"Symmetric" would be in my opinion either:
* forward is positive and backward is negative and the opposite sign is considered invalid or
* forward as well as backward accept either sign and always use the absolute value
(in both cases 0 represents the disabled state).

Reply:
>"Symmetric" would be in my opinion either:
>
> * forward is positive and backward is negative and the opposite sign is considered invalid or

Pushed 7da5d95f076892844d621e9932f88971b46c92b7 which implements this option